### PR TITLE
Restrict universes in comInductive

### LINKED ISSUE #7967: Universes of Inductives are not restricted
#### Version

coq master e25d69f

#### Description of the problem

```coq
Set Universe Polymorphism.
Inductive A : Set := B : ltac:(let y := constr:(Type) in exact nat) -> A.
Fail Check A@{}. (* Universe instance should have length 1 *)
```

`A` ends up with one universe even though it is not used anywhere.

### PR DESCRIPTION
**Pull request by jashug**
**Kind:** bug fix

Fixes #7967

- [x] Added / updated test-suite

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Do you want to do it for records too? (vernac/record.ml)

**Comment by jashug** [comment]
Given what I've just been working on, I'm much less familiar with `vernac/record.ml` than `vernac/comInductive.ml`. It does look like `Record` is also missing restriction, but I don't want to spend the time fixing it right now.

File: vernac/comInductive.ml [hunk 1]
```diff
@@ -379,6 +379,14 @@ let interp_mutual_inductive (paramsl,indl) notations cum poly prv finite =
...
   let constructors = List.map (fun (idl,cl,impsl) -> (idl,List.map nf cl,impsl)) constructors in
   let ctx_params = List.map Termops.(map_rel_decl (EConstr.to_constr sigma)) ctx_params in
+  (* Only keep useful universes *)
+  let merge_universes_of_constr c =
+    Univ.LSet.union (EConstr.universes_of_constr sigma (EConstr.of_constr c)) in
+  let uvars = Univ.LSet.empty in
+  let uvars = Context.Rel.(fold_outside (Declaration.fold_constr merge_universes_of_constr) ctx_params ~init:uvars) in
+  let uvars = List.fold_right merge_universes_of_constr arities uvars in
+  let uvars = List.fold_right (fun (_,ctypes,_) -> List.fold_right merge_universes_of_constr ctypes) constructors uvars in
+  let sigma = Evd.restrict_universe_context sigma uvars in
```
Comment (ejgallego) [inline vernac/comInductive.ml]: wow that looks like a very large amount of code to very inlined here, shouldn't it factored out elsewhere?

**Comment by jashug** [comment]
I pulled the universe restriction out into a helper function. I'm not sure this helper function is useful anywhere else, because the representation it works on (parameters, arities, constructors) is pretty ad-hoc.

**Comment by ejgallego** [comment]
Who's taking care of this?

**Comment by maximedenes** [comment]
It should be @mattam82.

**Comment by ejgallego** [comment]
> It should be @mattam82.

You are the secondary tho.

**Comment by maximedenes** [comment]
> You are the secondary tho.

Yeah but this is undocumented universe stuff (restriction). I'm not competent, sorry.

**Comment by ejgallego** [comment]
>

**Comment by ejgallego** [comment]
Let's try with @SkySkimmer

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Damn button not needing confirmation.

**Comment by SkySkimmer** [comment]
Why is coqbot adding needs: rebase to random PRs?

**Comment by Zimmi48** [comment]
@SkySkimmer Because the PR is not up-to-date with master. CI is not run anymore if at the time of pushing / opening, the PR is not up-to-date with the base branch.

**Comment by Zimmi48** [comment]
> That should probably be explained in contributing.md

Right. I intended to do this. Will do now.

**Comment by mattam82** [comment]
@SkySkimmer this looks ready, no?

**Comment by SkySkimmer** [comment]
I seem to have missed that I was supposed to be assigned ;)

**Pull request merged**